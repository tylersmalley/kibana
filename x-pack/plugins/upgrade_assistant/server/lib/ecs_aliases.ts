/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import _ from 'lodash';

const DEFAULT_LIMIT = 20;
const DEFAULT_BATCH_SIZE = 30;

export class EcsAliases {
  public static convertToPropertyPath(propertyPath: string) {
    return propertyPath.split('.').join('.properties.');
  }

  public static mappingFromFields(fields: any) {
    return {
      properties: Object.keys(fields).reduce((acc, key) => {
        _.set(acc, EcsAliases.convertToPropertyPath(fields[key]), {
          type: 'alias',
          path: key,
        });
        return acc;
      }, {}),
    };
  }
  constructor(
    boundCallWithRequest: any,
    registrations: any,
    batchSize?: number = DEFAULT_BATCH_SIZE
  ) {
    this.boundCallWithRequest = boundCallWithRequest;
    this.registrations = registrations;
    this.batchSize = batchSize;
  }

  public async fetch(offset: number = 0, limit: number = DEFAULT_LIMIT, indices?: string[]) {
    if (!indices) {
      indices = await this.fetchIndices();
    }

    const suggestions = [];
    let batch = [];
    let nextOffset: number;

    for (let i: number = offset; i <= indices.length; i++) {
      batch.push(indices[i]);

      if (batch.length === this.batchSize || i === indices.length) {
        const selectedMappings = await this.boundCallWithRequest('indices.getMapping', {
          index: batch.join(','),
        });

        batch = []; // reset

        for (const [index, { mappings }] of Object.entries(selectedMappings)) {
          if (suggestions.length >= limit) {
            nextOffset = indices.indexOf(_.last(suggestions).index) + 1;
            break;
          }

          const mappingTypes = Object.keys(mappings);
          const mappingType = mappingTypes[0];

          // we are only interested in mappings with a single type
          // as Beats have never had multiple types within an index
          if (mappingTypes.length !== 1) {
            return;
          }

          const type = mappingTypes[0];
          const properties = mappings[type].properties;

          const fields = this.registrations.reduce((acc, convert) => {
            const aliasedFields = _.pick(convert(properties), (value, key) => {
              // exclude if source property does not exist
              if (!_.has(mappings, EcsAliases.convertToPropertyPath(`${mappingType}.${key}`))) {
                return false;
              }

              // exclude if destination property already exists
              if (_.has(mappings, EcsAliases.convertToPropertyPath(`${mappingType}.${value}`))) {
                return false;
              }

              return true;
            });

            return _.merge({}, acc, aliasedFields);
          }, {});

          if (Object.keys(fields).length > 0) {
            suggestions.push({
              index,
              type,
              fields,
              mapping: EcsAliases.mappingFromFields(fields),
            });
          }
        }
      }
    }

    return { nextOffset, suggestions };
  }

  private async create(indices: string[]) {
    const { suggestions } = await this.fetch(0, DEFAULT_LIMIT, indices);

    const updates = suggestions.map(async ({ mapping, index, type }) => {
      const update = await this.boundCallWithRequest('indices.putMapping', {
        index,
        type,
        body: mapping,
      });

      return update.acknowledged === true;
    });

    return Promise.all(updates);
  }

  private async fetchIndices() {
    const indices = await this.boundCallWithRequest('cat.indices', {
      h: 'index',
      format: 'json',
      // pri: true, // TODO: can this be set?
    });

    return indices.reduce((acc: any, current: any) => {
      acc.push(current.index);
      return acc;
    }, []);
  }
}
