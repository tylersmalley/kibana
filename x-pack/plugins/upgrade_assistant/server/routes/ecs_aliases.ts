/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import Boom from 'boom';
import { Server } from 'hapi';
import Joi from 'joi';
import _ from 'lodash';

import { EcsAliases } from '../lib/ecs_aliases';

export function registerEcsAliasRoutes(server: Server) {
  const { callWithRequest } = server.plugins.elasticsearch.getCluster('data');

  server.route({
    path: '/api/upgrade_assistant/ecs_aliases',
    method: 'GET',
    options: {
      validate: {
        query: Joi.object()
          .keys({
            limit: Joi.number().default(20),
            offset: Joi.number().default(0),
            batch_size: Joi.number(),
          })
          .required(),
      },
    },
    async handler(request) {
      const boundCallWithRequest = _.partial(callWithRequest, request);
      const { batch_size: batchSize, limit, offset } = request.query;
      const { registrations } = server.plugins.upgrade_assistant.aliases;

      const aliases = new EcsAliases(boundCallWithRequest, registrations, batchSize);

      try {
        return await aliases.fetch(offset, limit);
      } catch (e) {
        return Boom.boomify(e, {
          statusCode: 500,
        });
      }
    },
  });

  server.route({
    path: '/api/upgrade_assistant/ecs_aliases/{indices}',
    method: 'PUT',
    options: {
      validate: {
        params: Joi.object()
          .keys({
            indices: Joi.string().required(),
          })
          .required(),
      },
    },
    async handler(request) {
      const boundCallWithRequest = _.partial(callWithRequest, request);
      const { registrations } = server.plugins.upgrade_assistant.aliases;

      const aliases = new EcsAliases(boundCallWithRequest, registrations);

      try {
        const resp = await aliases.create(request.params.indices.split(','));
        return resp;
      } catch (e) {
        return Boom.boomify(e, {
          statusCode: 500,
        });
      }
    },
  });
}
