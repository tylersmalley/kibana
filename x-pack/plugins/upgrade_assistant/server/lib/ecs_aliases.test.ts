/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import _ from 'lodash';

import fakeCatIndices from './__fixtures__/fake_cat_indices.json';
import fakeMappings from './__fixtures__/fake_mappings.json';
import { EcsAliases } from './ecs_aliases';

describe('EcsAliases.mappingFromFields', () => {
  it('converts single field', () => {
    const fields = { foo: 'bar' };
    expect(EcsAliases.mappingFromFields(fields)).toEqual({
      properties: {
        bar: {
          path: 'foo',
          type: 'alias',
        },
      },
    });
  });

  it('converts multiple nested fields', () => {
    const fields = { 'beat.hostname': 'agent.hostname', 'beat.name': 'agent.name' };
    expect(EcsAliases.mappingFromFields(fields)).toEqual({
      properties: {
        agent: {
          properties: {
            hostname: {
              path: 'beat.hostname',
              type: 'alias',
            },
            name: {
              path: 'beat.name',
              type: 'alias',
            },
          },
        },
      },
    });
  });
});

describe('EcsAliases.convertToPropertyPath', () => {
  it('ignores if top level', () => {
    expect(EcsAliases.convertToPropertyPath('foo')).toEqual('foo');
  });

  it('converts to property path', () => {
    expect(EcsAliases.convertToPropertyPath('foo.bar')).toEqual('foo.properties.bar');
  });

  it('converts deeply nested property path', () => {
    expect(EcsAliases.convertToPropertyPath('foo.bar.baz')).toEqual(
      'foo.properties.bar.properties.baz'
    );
  });
});

describe('EcsAliases', () => {
  const registrations = [
    () => {
      return {
        'old.foo': 'new.newFoo',
      };
    },
    () => {
      return {
        'old.bar': 'new.newBar',
        'old.baz': 'new.newBaz',
      };
    },
  ];

  const boundCallWithRequest = jest.fn().mockImplementation(async (api, { index }) => {
    if (api === 'cat.indices') {
      return Promise.resolve(fakeCatIndices);
    } else if (api === 'indices.getMapping') {
      return Promise.resolve(_.pick(fakeMappings, index.split(',')));
    } else {
      throw new Error(`Unexpected API call: ${api}`);
    }
  });

  describe('fetch', () => {
    it('loads indices', async () => {
      const aliases = new EcsAliases(boundCallWithRequest, registrations, 15);
      const fetch = await aliases.fetch(0, 10);

      expect(fetch).toMatchSnapshot();
    });

    it('loads partial page at offset', async () => {
      const limit = 10;
      const aliases = new EcsAliases(boundCallWithRequest, registrations, 15);
      const fetch = await aliases.fetch(13, limit);

      expect(fetch.nextOffset).toBe(undefined);
      expect(fetch.suggestions).toHaveLength(9);

      expect(fetch).toMatchSnapshot();
    });
  });
});
