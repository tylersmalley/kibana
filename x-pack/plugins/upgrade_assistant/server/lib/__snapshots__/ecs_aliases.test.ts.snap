// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcsAliases fetch loads indices 1`] = `
Object {
  "nextOffset": 13,
  "suggestions": Array [
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-001",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-002",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-003",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-004",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-005",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-006",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-007",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-008",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-009",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-012",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
  ],
}
`;

exports[`EcsAliases fetch loads partial page at offset 1`] = `
Object {
  "nextOffset": undefined,
  "suggestions": Array [
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-013",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-014",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-015",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-016",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-017",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-018",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.foo": "new.newFoo",
      },
      "index": "beat-019",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
      },
      "index": "beat-020",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
    Object {
      "fields": Object {
        "old.bar": "new.newBar",
        "old.baz": "new.newBaz",
        "old.foo": "new.newFoo",
      },
      "index": "beat-021",
      "mapping": Object {
        "properties": Object {
          "new": Object {
            "properties": Object {
              "newBar": Object {
                "path": "old.bar",
                "type": "alias",
              },
              "newBaz": Object {
                "path": "old.baz",
                "type": "alias",
              },
              "newFoo": Object {
                "path": "old.foo",
                "type": "alias",
              },
            },
          },
        },
      },
      "type": "_doc",
    },
  ],
}
`;
